### PR TITLE
HIVE-27309: Large number of partitions and small files causes OOM in …

### PR DESCRIPTION
**Pull request by difin**
…query coordinator.

### What changes were proposed in this pull request?

Removed serialized table from Iceberg split and instead passing it via config.

### Why are the changes needed?

"org.apache.iceberg.SerializableTable" is getting serialized in every split takes a hit with large number of small splits. E.g in the case provided in the ticket, i had 100,000+ splits which were grouped to 41 splits.

However, during serialization "Table" information is serialized and each entity is around 7 KB. With 100,000 entries it will be easily 700 MB where as AM size itself is 2 GB. 
When large number of nested partitions (with small files) are read, AM bails out with OOM. 

### Does this PR introduce _any_ user-facing change?

No

### Is the change a dependency upgrade?

No

### How was this patch tested?

Passing pre-commit testing, manual testing on internal cloud env of the query provided in the ticket.

### CODE REVIEW

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergInputFormat.java [hunk 1]
```diff
@@ -136,7 +139,10 @@ public InputSplit[] getSplits(JobConf job, int numSplits) throws IOException {
     if (filter != null) {
       job.set(InputFormatConfig.FILTER_EXPRESSION, SerializationUtil.serializeToBase64(filter));
     }
-
+    Table table = Optional
+            .ofNullable(HiveIcebergStorageHandler.table(job, job.get(InputFormatConfig.TABLE_IDENTIFIER)))
+            .orElseGet(() -> Catalogs.loadTable(job));
+    job.set(InputFormatConfig.SERIALIZED_TABLE, SerializationUtil.serializeToBase64(table));
```
[the remark this answers is not in the document]
Comment (difin) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergInputFormat.java]: done

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java [hunk 2]
```diff
@@ -201,6 +200,15 @@ public List<InputSplit> getSplits(JobContext context) {
         .ofNullable(HiveIcebergStorageHandler.table(conf, conf.get(InputFormatConfig.TABLE_IDENTIFIER)))
         .orElseGet(() -> Catalogs.loadTable(conf));
 
+    if (conf.get(InputFormatConfig.TABLE_IDENTIFIER) == null) {
+      conf.set(InputFormatConfig.TABLE_IDENTIFIER, table.name());
+    }
+
+    if (conf.get(InputFormatConfig.SERIALIZED_TABLE_PREFIX + conf.get(InputFormatConfig.TABLE_IDENTIFIER)) == null) {
+      conf.set(InputFormatConfig.SERIALIZED_TABLE_PREFIX + conf.get(InputFormatConfig.TABLE_IDENTIFIER),
```
Comment (zhangbutao) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:208]: IIUC,  every split has a `conf `and each `conf `would contain a srialized iceberg table. I want to know if the srialized iceberg table would make `conf `bigger? and also would cause AM oom in case of many splits?
Did we have a benchmark test to verify the validation of this change?

Comment (difin) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:208]: - Conf contained serialized Iceberg table even before these changes and all q-tests besides one were passing without adding serialized table to conf. So these changes actually do not increase memory usage anywhere, but just decrease, because table is removed from the splits.

- One q-test that was failing was show_partitions_test.q. It was failing because it is doing select from _table.**partitions**._ 

- There were 2 unit test classes (not q-tests) that were failing because serialized table wasn't in conf, but it is something that was present only in unit tests.

For these 2 edge cases, I added serialized table into conf in `IcebergInputFormat.getSplits()`.

Comment (zhangbutao) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:208]: I almost see what you mean. thx

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java [hunk 3]
```diff
@@ -201,6 +200,15 @@ public List<InputSplit> getSplits(JobContext context) {
         .ofNullable(HiveIcebergStorageHandler.table(conf, conf.get(InputFormatConfig.TABLE_IDENTIFIER)))
         .orElseGet(() -> Catalogs.loadTable(conf));
 
+    if (conf.get(InputFormatConfig.TABLE_IDENTIFIER) == null) {
+      conf.set(InputFormatConfig.TABLE_IDENTIFIER, table.name());
```
Comment (zhangbutao) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:204]: I think we do not need this param. We can just use` table.name()`,

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java [hunk 4]
```diff
@@ -296,7 +303,8 @@ public void initialize(InputSplit split, TaskAttemptContext newContext) {
       CombinedScanTask task = ((IcebergSplit) split).task();
       this.context = newContext;
       this.conf = newContext.getConfiguration();
-      this.table = ((IcebergSplit) split).table();
+      this.table = SerializationUtil.deserializeFromBase64(
+                conf.get(InputFormatConfig.SERIALIZED_TABLE_PREFIX + conf.get(InputFormatConfig.TABLE_IDENTIFIER)));
```
Comment (zhangbutao) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:307]:
```suggestion
                conf.get(InputFormatConfig.SERIALIZED_TABLE_PREFIX + table.name()));
```

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java (hunk 3 above)
Comment (deniskuzZ) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:204]: also we can skip the if check as well

Comment (zhangbutao) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:204]: yes, this is also what i mean ;)

**Review by deniskuzZ — APPROVED** [review]
LGTM +1

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java [hunk 5]
```diff
@@ -201,6 +200,11 @@ public List<InputSplit> getSplits(JobContext context) {
         .ofNullable(HiveIcebergStorageHandler.table(conf, conf.get(InputFormatConfig.TABLE_IDENTIFIER)))
         .orElseGet(() -> Catalogs.loadTable(conf));
 
+    conf.set(InputFormatConfig.TABLE_IDENTIFIER, table.name());
```
Comment (deniskuzZ) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:203]: could we do the conf setup when loading the table: 
````
Table table = Optional
  .ofNullable(HiveIcebergStorageHandler.table(conf, conf.get(InputFormatConfig.TABLE_IDENTIFIER)))
  .orElseGet(() -> {
        Table hadoopTable = Catalogs.loadTable(conf);
        conf.set(InputFormatConfig.TABLE_IDENTIFIER, hadoopTable.name());
        conf.set(InputFormatConfig.SERIALIZED_TABLE_PREFIX + hadoopTable.name(), SerializationUtil.serializeToBase64(hadoopTable));
        return hadoopTable;
  });
````
